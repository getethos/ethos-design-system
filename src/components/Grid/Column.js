import PropTypes from 'prop-types'
import React, { useEffect } from 'react'

import RefType from './ref'
import styles from './Column.module.css'

export const Column = React.memo((props) => {
  const inlineStyles = {
    flexBasis: props.flexBasis ? props.flexBasis : '1',
  }

  useEffect(() => {
    if (props.active) {
      props.columnRef.current.focus()
    }
  }, [props.active])

  const tabIndex = props.active ? 0 : -1

  /**
   * This allows a child of the column (e.g. an anchor) to handle
   * tabIndex and point columnRef to itself so it gets the focus.
   */
  const interactiveRender = (props) => {
    const klasses = props.header
      ? [props.className, styles.columnheader].join(' ')
      : props.className
    return (
      <div
        className={klasses}
        style={inlineStyles}
        data-label={props.name}
        role={props.header ? 'columnheader' : 'cell'}
      >
        {props.children(props.active, props.columnRef)}
      </div>
    )
  }

  /**
   * Normal column so we place the columnRef on it directly.
   */
  const normalRender = (props) => {
    const klasses = props.header
      ? [props.className, styles.columnheader].join(' ')
      : props.className
    return (
      <div
        className={klasses}
        style={inlineStyles}
        data-label={props.name}
        // eslint-disable-next-line
        {...!props.sortable && { ref: props.columnRef }}
        role={props.header ? 'columnheader' : 'cell'}
        tabIndex={tabIndex}
      >
        {props.children(props.active)}
      </div>
    )
  }

  return props.interactive ? interactiveRender(props) : normalRender(props)
})

Column.displayName = 'Column'
Column.propTypes = {
  name: PropTypes.string.isRequired,
  active: PropTypes.bool,
  sortable: PropTypes.bool,
  columnRef: RefType.isRequired,
  children: PropTypes.func.isRequired,
  className: PropTypes.string,
  flexBasis: PropTypes.string,
  header: PropTypes.bool,
  interactive: PropTypes.bool,
}

Column.defaultProps = {
  active: false,
  className: styles.column,
  header: false,
  interactive: false,
}
