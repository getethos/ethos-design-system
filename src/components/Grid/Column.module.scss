.column {
  display: flex;
  transition: border-color 0.25s ease-in-out, box-shadow 0.1s ease-in-out;
  padding: var(--Space-16) var(--Space-12);
  box-sizing: border-box;
}

.columnheader {
  text-transform: uppercase;
  font-size: 11px;
  line-height: 16px;
  font-weight: 500;
  letter-spacing: 0.1em;
}

.column > a,
.column > a:link,
.column > a:visited,
.column > a:focus,
.column > a:hover,
.column > a:active {
  color: inherit;
  text-decoration: none;
  cursor: pointer;
}
