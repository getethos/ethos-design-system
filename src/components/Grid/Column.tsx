import * as React from 'react'
import { useEffect } from 'react'
import styles from './Column.module.scss'
type ColumnProps = {
  name: string
  active?: boolean
  sortable?: boolean
  columnRef: any
  className?: string
  flexBasis?: string
  header?: boolean
  interactive?: boolean
  children(active: boolean, columnRef?: any): JSX.Element
}
export const Column: React.SFC<ColumnProps> = React.memo((props) => {
  const inlineStyles = {
    flexBasis: props.flexBasis ? props.flexBasis : '1',
  }
  useEffect(() => {
    if (props.active) {
      props.columnRef.current.focus()
    }
  }, [props.active])
  const tabIndex = props.active ? 0 : -1
  /**
   * This allows a child of the column (e.g. an anchor) to handle
   * tabIndex and point columnRef to itself so it gets the focus.
   */
  const interactiveRender = (props) => {
    const klasses = props.header
      ? [props.className, styles.columnheader].join(' ')
      : props.className
    return (
      <div
        className={klasses}
        style={inlineStyles}
        data-label={props.name}
        role={props.header ? 'columnheader' : 'cell'}
      >
        {props.children(props.active, props.columnRef)}
      </div>
    )
  }
  /**
   * Normal column so we place the columnRef on it directly.
   */
  const normalRender = (props) => {
    const klasses = props.header
      ? [props.className, styles.columnheader].join(' ')
      : props.className
    return (
      <div
        className={klasses}
        style={inlineStyles}
        data-label={props.name}
        // eslint-disable-next-line
        {...(!props.sortable && { ref: props.columnRef })}
        role={props.header ? 'columnheader' : 'cell'}
        tabIndex={tabIndex}
      >
        {props.children(props.active)}
      </div>
    )
  }
  return props.interactive ? interactiveRender(props) : normalRender(props)
})
Column.displayName = 'Column'
Column.defaultProps = {
  active: false,
  className: styles.column,
  header: false,
  interactive: false,
}
