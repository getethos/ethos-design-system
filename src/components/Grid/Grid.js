import PropTypes from 'prop-types'
import React, { Children } from 'react'

import { Position } from './Position'
import RefType from './ref'
import styles from './Grid.module.scss'

export const Grid = (props) => {
  const gridRefs = [...props.columnRefs, ...props.rowRefs]
  return (
    <Position className={props.className} refs={gridRefs} role="grid">
      {(positionX, positionY) =>
        Children.map(props.children, (row, index) => {
          // From https://bit.ly/2tq3FiS
          // ...`props.children` isn't the actual children; It is the descriptor
          // of the children. So you don't have actually anything to change; you
          // can't change any props, or edit any functionality; you can only read
          // from it. If you need to make any modifications you have to create new
          // elements using React.CloneElement.
          return React.cloneElement(row, {
            active: index === positionY,
            columnIndex: positionX,
          })
        })
      }
    </Position>
  )
}

Grid.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  // Grid's a 2d array or matrix, so we expect an array of arrays of refs
  rowRefs: PropTypes.arrayOf(PropTypes.arrayOf(RefType)).isRequired,
  columnRefs: PropTypes.arrayOf(PropTypes.arrayOf(RefType)).isRequired,
}

Grid.defaultProps = {
  className: styles.container,
}

Grid.displayName = 'Grid'
