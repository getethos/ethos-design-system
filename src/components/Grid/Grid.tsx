import * as React from 'react'
import styles from './Grid.module.scss'
import { Position } from './Position'
type GridProps = {
  className?: string
  rowRefs: any[][]
  columnRefs: any[][]
}
export const Grid: React.SFC<GridProps> = (props) => {
  const gridRefs = [...props.columnRefs, ...props.rowRefs]
  return (
    <Position className={props.className} refs={gridRefs} role="grid">
      {(positionX, positionY) =>
        // ethan - i dont understand the ts error message
        // @ts-ignore
        React.Children.map(
          props.children,
          (
            row: React.ReactElement<{ active: boolean; columnIndex: number }>,
            index
          ) => {
            // From https://bit.ly/2tq3FiS
            // ...`props.children` isn't the actual children; It is the descriptor
            // of the children. So you don't have actually anything to change; you
            // can't change any props, or edit any functionality; you can only read
            // from it. If you need to make any modifications you have to create new
            // elements using React.CloneElement.
            return React.cloneElement(row, {
              active: index === positionY,
              columnIndex: positionX,
            })
          }
        )
      }
    </Position>
  )
}
Grid.defaultProps = {
  className: styles.container,
}
Grid.displayName = 'Grid'
