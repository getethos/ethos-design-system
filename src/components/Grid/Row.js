import PropTypes from 'prop-types'
import React from 'react'
import styles from './Row.module.scss'

export const Row = React.memo((props) => {
  return (
    <div className={props.className} role="row">
      {React.Children.map(props.children, (column, index) =>
        // From https://bit.ly/2tq3FiS
        // ...`props.children` isn't the actual children; It is the descriptor
        // of the children. So you don't have actually anything to change; you
        // can't change any props, or edit any functionality; you can only read
        // from it. If you need to make any modifications you have to create new
        // elements using React.CloneElement.
        React.cloneElement(column, {
          active: props.active && index === props.columnIndex,
        })
      )}
    </div>
  )
})

Row.displayName = 'Row'

Row.propTypes = {
  active: PropTypes.bool,
  columnIndex: PropTypes.number,
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
}

Row.defaultProps = {
  active: false,
  columnIndex: -1,
  className: styles.row,
}
