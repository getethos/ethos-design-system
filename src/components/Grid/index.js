export { Grid } from './Grid.js'
export { Row } from './Row.js'
export { Column } from './Column.js'
export { useGridSorting } from './useGridSorting.js'
