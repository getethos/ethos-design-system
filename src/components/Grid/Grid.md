### Accessible React Grid
This `Grid` follows the [wai-aria-practices-1.1](https://www.w3.org/TR/wai-aria-practices-1.1/#grid) specifications for data grids. It's influenced by [juanca's](https://github.com/juanca) original [react-aria-components](https://github.com/juanca/react-aria-components) data-grid component in the way the keyboard navigation and roles are assigned, but otherwise is a complete rewrite with the following added features:

- The class-based lifecycle components were all converted to use functional hook-based components
- Simpler API
- Boilerplate default styles
- Sort by column
- Responsive stacking at mobile and lower with column label on left and values on the right

### Aria Support in Screen Readers

While support from screenreaders for aria roles is not yet perfect, it's on an [upward trend](https://www.powermapper.com/tests/screen-readers/aria/) so it would seem that having a data grid that complies is a worthwhile endeavor!

### Usage

Place focus anywhere just above the data grid and then tab. You've put focus on the grid. Now, use your arrow keys to navigate within the grid. On a Mac, `fn-rightarrow` simulates a _Page Right_, `fn-downarrow` a _Page Down_ and so on.


```jsx
import React, { useState } from 'react'
import uuidv4 from 'uuid/v4'
import { FaHamburger } from 'react-icons/fa'
import { Grid } from './Grid.js'
import { Row } from './Row.js'
import { Column } from './Column.js'
import { useGridSorting } from './useGridSorting.js'
import styles from './grid-example.module.css'

const columns = [
  {
    name: 'Description',
    interactive: true,
    sortable: true /* will use default sort function */,
    flexBasis: '40%',
  },
  {
    name: 'Type',
    flexBasis: '30%',
  },
  {
    name: 'Date',
    flexBasis: '15%',
    sortable: true,
    // Custom sort function example
    sortFn: (a, b) => {
      if (a > b) {
        return 1
      } else if (a < b) {
        return -1
      }
      return 0
    },
  },
  {
    name: 'Cost',
    flexBasis: '15%',
    interactive: true,
    sortable: true /* will use default sort function */,
  },
]
const rows = [
  {
    Description: 'Learn Typescript',
    Type: 'Education',
    Date: '2020-01-15',
    Cost: '$100.00',
  },
  {
    Description: 'Las Vegas Trip',
    Type: 'Travel',
    Date: '2020-01-04',
    Cost: '$999.99',
  },
  {
    Description: 'Coffee and Bagel',
    Type: 'Food',
    Date: '2020-02-22',
    Cost: '$11.00',
  },
  {
    Description: 'Guitar Lessons',
    Type: 'Education',
    Date: '2018-03-11',
    Cost: '$300.00',
  },
  {
    Description: 'Haircut',
    Type: 'Grooming',
    Date: '2019-12-19',
    Cost: '$25.00',
  },
]

function GridExample() {
  const {
    rowsRefs,
    columnRefs,
    sortedRows,
    compareBy,
    updateRowsRefs,
    getSortIcon,
  } = useGridSorting(rows, columns)

  const sortBy = (ev) => {
    ev.preventDefault()
    const key = ev.currentTarget.text
    let rowsCopy = sortedRows
    let sortFunction
    // See if wee have a custom sort function. If `sortFunction` is
    // `undefined`, `compareBy` will fallback to its own default sort
    const idx = columns.findIndex((c) => c.name === key)
    if (columns[idx].sortFn) {
      sortFunction = columns[idx].sortFn
    }
    rowsCopy.sort(compareBy(key, sortFunction))
    updateRowsRefs(rowsCopy)
  }

  return (
    columns.length > 0 && (
      <Grid rowRefs={rowsRefs} columnRefs={columnRefs}>
        <Row key="headers" columnRefs={columnRefs[0]}>
          {columns.map((col, x) => {
            if (col.sortable) {
              return (
                <Column
                  name={col.name}
                  key={uuidv4()}
                  columnRef={columnRefs[0][x]}
                  flexBasis={col.flexBasis}
                  header
                  interactive
                  sortable
                >
                  {(active, columnRef) => (
                    <a
                      href="./#"
                      onClick={sortBy}
                      tabIndex={active ? 0 : -1}
                      className={styles.iconContainer}
                      ref={columnRef}
                    >
                      {col.name}
                      {getSortIcon(col.name)}
                    </a>
                  )}
                </Column>
              )
            }
            return (
              <Column
                name={col.name}
                key={uuidv4()}
                columnRef={columnRefs[0][x]}
                flexBasis={col.flexBasis}
                header
              >
                {(active) => (
                  <span className={active ? 'active' : undefined}>
                    {col.name}
                  </span>
                )}
              </Column>
            )
          })}
        </Row>
        {sortedRows.map((row, y) => (
          <Row key={uuidv4()} columnRefs={rowsRefs[y]}>
            {columns.map((col, x) => {
              if (col.interactive) {
                return (
                  <Column
                    name={col.name}
                    key={uuidv4()}
                    columnRef={rowsRefs[y][x]}
                    flexBasis={col.flexBasis}
                    interactive
                  >
                    {(active, columnRef) => (
                      <a href="./#" tabIndex={active ? 0 : -1} ref={columnRef}>
                        {row[col.name]}
                      </a>
                    )}
                  </Column>
                )
              }
              return (
                <Column
                  name={col.name}
                  key={uuidv4()}
                  columnRef={rowsRefs[y][x]}
                  flexBasis={col.flexBasis}
                >
                  {() => (
                    <div className={styles.iconContainer}>
                      {row[col.name]}{' '}
                      {col.name === 'Type' && row[col.name] === 'Food' && (
                        <FaHamburger className={styles.icon} />
                      )}
                    </div>
                  )}
                </Column>
              )
            })}
          </Row>
        ))}
      </Grid>
    )
  )
}

<GridExample />
```

You can pass `Small` or `Large` to the `Row` component to get a compressed or enlarged grid.
Here's an example of using `Small` and `Large` on jut the row with _Guitar Lessons_:

```jsx
import React, { useState } from 'react'
import uuidv4 from 'uuid/v4'
import { FaHamburger } from 'react-icons/fa'
import { Grid } from './Grid.js'
import { Row } from './Row.js'
import { Column } from './Column.js'
import { useGridSorting } from './useGridSorting.js'
import styles from './grid-example.module.css'

const columns = [
  {
    name: 'Description',
    interactive: true,
    sortable: true /* will use default sort function */,
    flexBasis: '40%',
  },
  {
    name: 'Type',
    flexBasis: '30%',
  },
  {
    name: 'Date',
    flexBasis: '15%',
    sortable: true,
    // Custom sort function example
    sortFn: (a, b) => {
      if (a > b) {
        return 1
      } else if (a < b) {
        return -1
      }
      return 0
    },
  },
  {
    name: 'Cost',
    flexBasis: '15%',
    interactive: true,
    sortable: true /* will use default sort function */,
  },
]
const rows = [
  {
    Description: 'Learn Typescript',
    Type: 'Education',
    Date: '2020-01-15',
    Cost: '$100.00',
  },
  {
    Description: 'Las Vegas Trip',
    Type: 'Travel',
    Date: '2020-01-04',
    Cost: '$999.99',
  },
  {
    Description: 'Coffee and Bagel',
    Type: 'Food',
    Date: '2020-02-22',
    Cost: '$11.00',
  },
  {
    Description: 'Guitar Lessons',
    Type: 'Education',
    Date: '2018-03-11',
    Cost: '$300.00',
  },
  {
    Description: 'Haircut',
    Type: 'Grooming',
    Date: '2019-12-19',
    Cost: '$25.00',
  },
]

function GridSmall() {
  const {
    rowsRefs,
    columnRefs,
    sortedRows,
    compareBy,
    updateRowsRefs,
    getSortIcon,
  } = useGridSorting(rows, columns)

  const sortBy = (ev) => {
    ev.preventDefault()
    const key = ev.currentTarget.text
    let rowsCopy = sortedRows
    let sortFunction
    // See if wee have a custom sort function. If `sortFunction` is
    // `undefined`, `compareBy` will fallback to its own default sort
    const idx = columns.findIndex((c) => c.name === key)
    if (columns[idx].sortFn) {
      sortFunction = columns[idx].sortFn
    }
    rowsCopy.sort(compareBy(key, sortFunction))
    updateRowsRefs(rowsCopy)
  }

  return (
    columns.length > 0 && (
      <Grid rowRefs={rowsRefs} columnRefs={columnRefs}>
        <Row size="Small" key="headers" columnRefs={columnRefs[0]}>
          {columns.map((col, x) => {
            if (col.sortable) {
              return (
                <Column
                  name={col.name}
                  key={uuidv4()}
                  columnRef={columnRefs[0][x]}
                  flexBasis={col.flexBasis}
                  header
                  interactive
                  sortable
                >
                  {(active, columnRef) => (
                    <a
                      href="./#"
                      onClick={sortBy}
                      tabIndex={active ? 0 : -1}
                      className={styles.iconContainer}
                      ref={columnRef}
                    >
                      {col.name}
                      {getSortIcon(col.name)}
                    </a>
                  )}
                </Column>
              )
            }
            return (
              <Column
                name={col.name}
                key={uuidv4()}
                columnRef={columnRefs[0][x]}
                flexBasis={col.flexBasis}
                header
              >
                {(active) => (
                  <span className={active ? 'active' : undefined}>
                    {col.name}
                  </span>
                )}
              </Column>
            )
          })}
        </Row>
        {sortedRows.map((row, y) => (
          <Row size={y === 3 ? "Large" : "Small"} key={uuidv4()} columnRefs={rowsRefs[y]}>
            {columns.map((col, x) => {
              if (col.interactive) {
                return (
                  <Column
                    name={col.name}
                    key={uuidv4()}
                    columnRef={rowsRefs[y][x]}
                    flexBasis={col.flexBasis}
                    interactive
                  >
                    {(active, columnRef) => (
                      <a href="./#" tabIndex={active ? 0 : -1} ref={columnRef}>
                        {row[col.name]}
                      </a>
                    )}
                  </Column>
                )
              }
              return (
                <Column
                  name={col.name}
                  key={uuidv4()}
                  columnRef={rowsRefs[y][x]}
                  flexBasis={col.flexBasis}
                >
                  {() => (
                    <div className={styles.iconContainer}>
                      {row[col.name]}{' '}
                      {col.name === 'Type' && row[col.name] === 'Food' && (
                        <FaHamburger className={styles.icon} />
                      )}
                    </div>
                  )}
                </Column>
              )
            })}
          </Row>
        ))}
      </Grid>
    )
  )
}

<GridSmall />

```
