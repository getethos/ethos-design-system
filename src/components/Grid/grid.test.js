import React from 'react'
import renderer from 'react-test-renderer'

import Grid from './grid.js'
import Cell from './column.js'
import Row from './row.js'

test('renders the default state', () => {
  const grid = renderer
    .create(
      <Grid rowRefs={[]} columnRefs={[]}>
        <div>Foo</div>
        <div>Bar</div>
        <div>Baz</div>
      </Grid>
    )
    .toJSON()
  expect(grid).toMatchSnapshot()
})
