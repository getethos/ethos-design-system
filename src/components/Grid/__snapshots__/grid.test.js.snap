// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the default state 1`] = `
<div
  className="container"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="grid"
  tabIndex={0}
>
  <div
    active={false}
    columnIndex={-1}
  >
    Foo
  </div>
  <div
    active={false}
    columnIndex={-1}
  >
    Bar
  </div>
  <div
    active={false}
    columnIndex={-1}
  >
    Baz
  </div>
</div>
`;
